### ISSUE TITLE
"There are unsaved changes, are you sure you want to leave the page?"

### ISSUE THREAD

**Issue by coreymckrill** [WordPress/revisions-extended#49]
The dialog in the screenshot displays after the revision is created because of unsaved changes in the original document.

### Screenshot
![](https://d.pr/i/jma6sc.png)

**Comment by StevenDufresne** [comment]
Preliminary Debugging:

**Facts**:
- If any change is made to the post/page's content, refreshing the page will trigger the popup.
  - Even if the changes are reversed using the undo function, the dialog still appears.
- The dialog leaves when the `savePost` function is called.
  - Switching to draft
  - Publishing/Updating.

**Comment by ryelle** [comment]
I _think_ this comes from the [UnsavedChangesWarning](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/unsaved-changes-warning/index.js) component, which checks for ["dirty" records (unsaved edits)](https://github.com/WordPress/gutenberg/blob/98e96caf8b4668839a4a2a14ade1b37fe272bc1f/packages/core-data/src/selectors.js#L279). Maybe you could clear the edits on the post when the revision is saved? That might visually change the content in the editor, though.

Basically, I think you want [`getEntityRecordNonTransientEdits`](https://github.com/WordPress/gutenberg/blob/98e96caf8b4668839a4a2a14ade1b37fe272bc1f/packages/core-data/src/selectors.js#L351) to return an empty object.

You can debug some of this by calling the selectors in the console, like:

`wp.data.select('core'). getEntityRecordNonTransientEdits( 'postType', 'post', postId )`

I think you could fully disable this by removing the `beforeunload` listener, but that's a bit  💥

**Comment by StevenDufresne** [comment]
Thanks!

> I think you want [`getEntityRecordNonTransientEdits`](https://github.com/WordPress/gutenberg/blob/98e96caf8b4668839a4a2a14ade1b37fe272bc1f/packages/core-data/src/selectors.js#L351) to return an empty object.

Right, that makes sense, I tried updating the post using the original object, something like `editPost( wp.[...].currentPost() )`  and `savePost()` (which is how the switch to draft works) but that would still grab the edits. I'll see if we can delete all the entries.  

> I think you could fully disable this by removing the `beforeunload` listener, but that's a bit 💥

I tried returning `undefined` here as in the MDN docs but it didn't work, the browser would call the `beforeunload` event but still show the popup; before I knew about the `UnsavedChangesWarning` component. But the dialog never matches the component's copy so I think they are different somehow.

**Comment by ryelle** [comment]
> But the dialog never matches the component's copy so I think they are different somehow.

I _think_ I remember reading something about how some browsers don't let you control this copy… aha, found it:

> Starting with Firefox 44, Chrome 51, Opera 38, and Safari 9.1, a generic string not under the control of the webpage will be shown instead of the returned string.

[source](https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload#browser_compatibility)

**Comment by StevenDufresne** [comment]
@ryelle I think I have a fix: #57. 

Mind taking a look?
